nsxt/resource_nsxt_node_user.go

```diff
@@ -0,0 +1,240 @@
+	}
+
+	// Password not return on GET
+	d.Set("full_name", user.FullName)
+	d.Set("last_password_change", user.LastPasswordChange)
+	d.Set("password_change_frequency", user.PasswordChangeFrequency)
+	d.Set("password_change_warning", user.PasswordChangeWarning)
+	d.Set("password_reset_required", user.PasswordResetRequired)
+	d.Set("status", user.Status)
+	d.Set("user_id", user.Userid)
+	d.Set("username", user.Username)
+	if *user.Status == nsxModel.NodeUserProperties_STATUS_NOT_ACTIVATED {
+		d.Set("active", false)
+	} else {
+		d.Set("active", true)
+	}
+	return nil
+}
+
+func resourceNsxtNodeUserUpdate(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := node.NewUsersClient(connector)
+	id := d.Id()
+	if id == "" {
+		return fmt.Errorf("error obtaining logical object id")
+	}
+	password := d.Get("password").(string)
+	active := d.Get("active").(bool)
+	status := d.Get("status").(string)
+
+	// Handle user status change first
+	// Password reset can be achieved by deactivating then re-activate the account
+	if status == nsxModel.NodeUserProperties_STATUS_NOT_ACTIVATED && active {
+		if len(password) == 0 {
+			return fmt.Errorf("must specify password to activate Nsxt Node user %s", id)
+		}
+		_, err := client.Activate(id, nsxModel.NodeUserPasswordProperty{Password: &password})
+		if err != nil {
+			return fmt.Errorf("failed to activate Nsxt Node user %s: %s", id, err)
+		}
+	} else if status != nsxModel.NodeUserProperties_STATUS_NOT_ACTIVATED && !active {
+		_, err := client.Deactivate(id)
+		if err != nil {
+			return fmt.Errorf("failed to deactivate Nsxt Node user %s: %s", id, err)
+		}
+	}
+
+	// Update other user properties
+	fullName := d.Get("full_name").(string)
+	passwordChangeFrequency := int64(d.Get("password_change_frequency").(int))
+	passwordChangeWarning := int64(d.Get("password_change_warning").(int))
+	username := d.Get("username").(string)
+	userProp := nsxModel.NodeUserProperties{
+		FullName:                &fullName,
+		PasswordChangeFrequency: &passwordChangeFrequency,
+		PasswordChangeWarning:   &passwordChangeWarning,
+		Username:                &username,
+	}
+	_, err := client.Update(id, userProp)
+	if err != nil {
+		return handleUpdateError("User", id, err)
+	}
+	return resourceNsxtNodeUserRead(d, m)
+}
+
+func resourceNsxtNodeUserDelete(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := node.NewUsersClient(connector)
+	id := d.Id()
+	if id == "" {
+		return fmt.Errorf("error obtaining logical object id")
+	}
+
+	// Workaround a bug where the delete response body is marked as non-json content type by NSX,
+	// triggering SDK to raise it as internal server error.
+	// Error Data is not complying with ApiErrorBindingType, making the conversion result empty.
+	// A second deletion attempt should be deemed as NotFound, which is then treated as successful.
+	var err error
+	for i := 0; i < 2; i++ {
+		err = client.Delete(id)
```

Comment:
does the first delete always return error?

Reply:
Essentially, the first DELETE call actually went through, with HTTP status 200. It's just the returned HTTP header `content-type`: `text/html; charset=UTF-8` would always trigger sdk to report it back as internal server error, and such error type casting made checking status code very difficult on our end.

The second call will make NSX return 404, which is ignored on delete calls.

This issue still exists on NSX 4.1.2 rtqa.